lib/Transforms/ShapeInference/ShapeInference.cpp

```diff
@@ -0,0 +1,406 @@
+                      << input << " with type of successor operand: " << operand
+                      << "\n");
+                  input.setType(operand.getType());
+                }
+              }
+              // While some ops might have both RegionBranchOpInterface and
+              // other interfaces, such as InferTypeOpInterface, we cannot rely
+              // on those since the body and terminators have not yet been
+              // updated and the op is therefore potentially invalid.
+              // -> return early and skip other interface handling code
+              return WalkResult::advance();
+            }
+
+            // We focus on operations with return type inference,
+            // as these will likely complain if their operands' types change
+            // but their return type hasn't been updated.
+            // Technically, other ops could also cause issues,
+            // For example, an op could allow only dynamically shaped tensors,
+            // which means it would fail to verify after shape inference,
+            // but failing on those is probably the right thing to do for the
+            // compiler
+            if (auto inferOp = dyn_cast<InferTypeOpInterface>(op)) {
+              LLVM_DEBUG(llvm::dbgs() << "ShapeInference:\t\t\tcurrent op has "
+                                         "InferTypeOpInterface.\n");
+              SmallVector<Type> inferredReturnTypes;
+              if (failed(inferOp.inferReturnTypes(
+                      rewriter.getContext(), op->getLoc(), op->getOperands(),
+                      op->getAttrDictionary(), op->getPropertiesStorage(),
+                      op->getRegions(), inferredReturnTypes))) {
+                emitError(op->getLoc(), "Failed to infer return types.");
+                return WalkResult::interrupt();
+              }
+
+              assert(inferredReturnTypes.size() == op->getNumResults() &&
+                     "Number of inferred return types must match the number of "
+                     "results");
+
+              for (size_t i = 0; i < inferredReturnTypes.size(); ++i) {
+                if (inferredReturnTypes[i] == op->getResult(i).getType())
+                  continue;
+                op->getResult(i).setType(inferredReturnTypes[i]);
+              }
+
+              if (failed(verify(op))) {
+                emitError(op->getLoc(),
+                          "Operation failed to verify with its newly inferred "
+                          "return type(s) after its operands' types were "
+                          "updated during shape inference.");
+                return WalkResult::interrupt();
+              }
+            }
+
+            // If an operation is a (RegionBranch) terminator, we need to update
+            // the return types of the results of the parent operation to match
+            // the types of the terminators operands.
+            if (auto terminator =
```

Comment:
My only concern now is readability: this function is massive and some parts, like the handlers for each interface type, seem like they can be split into independent functions without too much hassle.